File: lld/problems/category_b_machine_coding/ingolang/in_memory_job_scheduler_go/dao/jobDao.go
package dao

import (
	"in_memory_job_scheduler_go/models"
	"sync"
	"time"
)

type JobDao struct {
	mu                sync.RWMutex
	immediateJobStore map[int]*models.ImmediateJob
	scheduleJobStore  map[int]*models.ScheduleJob
	executionJob      map[int]*models.ExecutionJob
}

func NewJobDao() *JobDao {
	return &JobDao{
		immediateJobStore: make(map[int]*models.ImmediateJob),
		scheduleJobStore:  make(map[int]*models.ScheduleJob),
		executionJob:      make(map[int]*models.ExecutionJob),
	}
}

func (d *JobDao) CreateImmediateJob(job *models.ImmediateJob) *models.ImmediateJob {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.immediateJobStore[job.Id] = job
	return job
}

func (d *JobDao) CreateExecutionJob(job *models.ExecutionJob) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.executionJob[job.Id] = job
}

func (d *JobDao) CreateSchedule(job *models.ScheduleJob) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.scheduleJobStore[job.Id] = job
}

func (d *JobDao) GetJob(id int, jobType models.JobType) models.JobI {
	d.mu.RLock()
	defer d.mu.RUnlock()
	switch jobType {
	case models.RECURRING:
		return d.scheduleJobStore[id]
	case models.NON_RECURRING:
		return d.immediateJobStore[id]
	}
	return nil
}

func (d *JobDao) GetExJobRange(st, et time.Time) []*models.ExecutionJob {
	d.mu.RLock()
	defer d.mu.RUnlock()
	var jobs []*models.ExecutionJob
	for _, ej := range d.executionJob {
		if ej.Status == models.CREATED && !ej.Ts.Before(st) && !ej.Ts.After(et) {
			jobs = append(jobs, ej)
		}
	}
	return jobs
}

func (d *JobDao) GetScJobRange(st, et time.Time) []*models.ScheduleJob {
	d.mu.RLock()
	defer d.mu.RUnlock()
	var jobs []*models.ScheduleJob
	for _, sj := range d.scheduleJobStore {
		if !sj.NextTs.Before(st) && !sj.NextTs.After(et) {
			jobs = append(jobs, sj)
		}
	}
	return jobs
}

func (d *JobDao) UpdateExJob(ej *models.ExecutionJob) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.executionJob[ej.Id] = ej
}

func (d *JobDao) UpdateScJob(sj *models.ScheduleJob) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.scheduleJobStore[sj.Id] = sj
}

File: lld/problems/category_b_machine_coding/incpp/in_memory_job_scheduler/dao/jobDao.cpp
#include<map> 
#include<iostream> 
#include"../models/job.cpp"
#include"../util/utils.cpp"
using namespace std; 

class JobDao {
    public:  // taken decision after writing few method , sicne its timetaking. we can do this in MC.
    map<int, ImmediateJob > immediateJobStore; 
    map<int, ScheduleJob > scheduleJobStore; 
    map<int, ExecutionJob  > executionJob; 
    
    JobDao(){};

    ImmediateJob createImmediateJob(ImmediateJob job, JobType type) {
        job.setJobId(Counter::getCount());
        this-> immediateJobStore[job.getJobId()] = job;
        return job;
    }
    void createExecutionJob(ExecutionJob job) {
        this-> executionJob[job.getJobId()] = job;
    }
    void createSchedule(ScheduleJob job) {
        this-> scheduleJobStore[job.getJobId()] = job;
    }
    JobI* getJob(int id, JobType type) {
        switch (type)
        {
        case JobType::RECURRING:
            return &this-> scheduleJobStore[id];
            break;
        case JobType::NON_RECURRING:
            return &this-> immediateJobStore[id];
            break;
        default:
            break;
        }
        
    }
    vector<ExecutionJob> getExJobRange(chrono::steady_clock::time_point st , chrono::steady_clock::time_point et) {
        vector<ExecutionJob> jobs; 
        for(auto ej : this -> executionJob) {
            if(ej.second.status == JobStatus::CREATED && ej.second.ts >= st && ej.second.ts <= et) {
                    jobs.push_back(ej.second);
            }
        }
        return jobs;
    }
    vector<ScheduleJob> getScJobRange(chrono::steady_clock::time_point st , chrono::steady_clock::time_point et) {
        vector<ScheduleJob> jobs; 
        for(auto ej : this -> scheduleJobStore) {
            if( ej.second.nextTs >= st && ej.second.nextTs <= et) {
                    jobs.push_back(ej.second);
            }
        }
        cout<<jobs.size()<<endl;
        return jobs;
    }
    void updateExJob(ExecutionJob ej) {
        this -> executionJob[ej.id] = ej;
    }
    void updateScJob(ScheduleJob sj) {
        this -> scheduleJobStore[sj.id] = sj;
    }
};

